File: plot.py
import numpy as np
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
import pandas as pd
import os
import json

def find_json(name):
    benchmark_path = os.path.join('.benchmarks', os.listdir('.benchmarks')[0])
    list = []
    for each in os.listdir(benchmark_path):
        if name in each:
            list.append(each)
    return os.path.join(benchmark_path, list[-1])

def wash_benchmark_data(name, labels):
    with open(find_json(name)) as f:
        data = json.load(f)

    cols = [each['params']['nqubits'] for each in data['benchmarks'] if each['group']==labels[0]]
    dd = {}
    dd['nqubits'] = cols
    for lb in labels:
        time_data = [each['stats']['min']*1e9
            for each in data['benchmarks'] if each['group']==lb]
        if len(time_data) is not len(cols):
            time_data.append([float('inf') for _ in range(len(cols) - len(time_data) + 1)])

        dd[lb] = time_data
    return pd.DataFrame(data=dd)

gate_data = {}
gate_set = ['X', 'H', 'T', 'CNOT', 'Toffoli']
packages = ['projectq', 'qiskit', 'cirq', 'quest', 'yao']

for each_package in packages:
    if each_package == 'yao':
        gate_data['yao'] = pd.read_csv('yao.csv')
    else:
        gate_data[each_package] = wash_benchmark_data(each_package, gate_set)


def subplot_dataset(dataset, ax, gate):
    ls = [ax.semilogy(dataset[each]["nqubits"], dataset[each][gate], '-o', markersize=3) for each in packages]
    return ls

def subplot_gate(ax, gate):
    ls = subplot_dataset(gate_data, ax, gate)
    ax.set(title=' '.join([gate, 'gate']), xlabel="nqubits", ylabel="ns")
    return ls

fig, ((ax1, ax2), (ax3, ax4)) = plt.subplots(2, 2, figsize=(10, 8), sharey=True)

ls = subplot_gate(ax1, 'X')
subplot_gate(ax2, 'H')
subplot_gate(ax3, 'CNOT')
subplot_gate(ax4, 'Toffoli')

lgd = fig.legend(
    ls,
    labels=packages,
    loc="upper right",
    borderaxespad=0.1,
    bbox_to_anchor=(1.09, 0.9))

plt.tight_layout()
plt.savefig('gates.png', bbox_extra_artists=(lgd,), bbox_inches='tight')

circuit_data = {}
for each_package in packages:
    if each_package == 'yao':
        circuit_data['yao'] = pd.read_csv('yao_qcbm.csv')
    elif each_package == 'qiskit':
        continue # skip qiskit, since not supported
    else:
        circuit_data[each_package] = wash_benchmark_data(each_package, ['QCBM'])

packages.append('pennylane')
circuit_data['pennylane'] = wash_benchmark_data('pennylane', ['QCBM'])
packages = [each for each in packages if each != 'qiskit']

fig = plt.figure(figsize=(8, 6))
ax = plt.subplot(111)
ls = subplot_dataset(circuit_data, ax, 'QCBM')

packages.append('yao (cuda)')
ls.append(ax.semilogy(circuit_data['yao']["nqubits"], circuit_data['yao']["QCBM_cuda"], '-o', markersize=3))
ax.set(title="Parameterized Circuit", xlabel="nqubits", ylabel="ns")

lgd = ax.legend(
    ls,
    labels=packages,
    loc="upper right",
    borderaxespad=0.1,
    bbox_to_anchor=(1.2, 0.9))
plt.savefig('pcircuit.png', bbox_extra_artists=(lgd,), bbox_inches='tight')

df_yao = pd.read_csv('yao_qcbm_batch.csv')
fig = plt.figure(figsize=(8, 6))
ax = plt.subplot(111)
l1 = ax.semilogy(df_yao["nqubits"], df_yao["QCBM_batch"], '-o', markersize=3)
l2 = ax.semilogy(df_yao["nqubits"], df_yao["QCBM_cuda_batch"], '-o', markersize=3)
ax.set(title="Batched Parameterized Circuit", xlabel="nqubits", ylabel="ns")
lgd = ax.legend(
    [l1, l2],
    labels=["CPU", "CUDA"],
    loc="upper right",
    borderaxespad=0.1,
    bbox_to_anchor=(1.2, 0.9))
plt.savefig('pcircuit_batch.png', bbox_extra_artists=(lgd,), bbox_inches='tight')

File: plot.jl
using JSON, DataFrames, CSV, Plots, JLD2

file = JSON.parsefile("cirq/.benchmarks/Darwin-CPython-3.7-64bit/0001_0246ecb3dfce8dae170ccdf222c115ce318fcdb9_20190702_184140_uncommited-changes.json")

qcbm_benchmark = Dict()
qcbm_benchmark["cirq"] = [each["stats"]["min"] * 1e9 for each in file["benchmarks"] if each["group"] == "QCBM"]

benchmarks_cirq = Dict()
benchmarks_cirq["X"] = [each["stats"]["min"] * 1e9 for each in file["benchmarks"] if each["group"] == "X"]
benchmarks_cirq["H"] = [each["stats"]["min"] * 1e9 for each in file["benchmarks"] if each["group"] == "H"]
benchmarks_cirq["T"] = [each["stats"]["min"] * 1e9 for each in file["benchmarks"] if each["group"] == "T"]
benchmarks_cirq["CNOT"] = [each["stats"]["min"] * 1e9 for each in file["benchmarks"] if each["group"] == "CNOT"]
benchmarks_cirq["Toffoli"] = [each["stats"]["min"] * 1e9 for each in file["benchmarks"] if each["group"] == "Toffoli"]

data = DataFrame(nqubits=4:25,
    CirqX=benchmarks_cirq["X"],
    CirqH=benchmarks_cirq["H"],
    CirqT=benchmarks_cirq["T"],
    CirqCNOT=benchmarks_cirq["CNOT"],
    CirqToffoli=benchmarks_cirq["Toffoli"])

CSV.write("cirq.csv", data)

file = JSON.parsefile("qiskit/.benchmarks/Darwin-CPython-3.7-64bit/0001_e5b779f8cec7ee91655814cb360026fad83667bd_20190703_201317_uncommited-changes.json")

qcbm_benchmark["qiskit"] = [each["stats"]["min"] * 1e9 for each in file["benchmarks"] if each["group"] == "QCBM"]

benchmarks_qiskit = Dict()
benchmarks_qiskit["X"] = [each["stats"]["min"] * 1e9 for each in file["benchmarks"] if each["group"] == "X"]
benchmarks_qiskit["H"] = [each["stats"]["min"] * 1e9 for each in file["benchmarks"] if each["group"] == "H"]
benchmarks_qiskit["T"] = [each["stats"]["min"] * 1e9 for each in file["benchmarks"] if each["group"] == "T"]
benchmarks_qiskit["CNOT"] = [each["stats"]["min"] * 1e9 for each in file["benchmarks"] if each["group"] == "CNOT"]
benchmarks_qiskit["Toffoli"] = [each["stats"]["min"] * 1e9 for each in file["benchmarks"] if each["group"] == "Toffoli"]

data = DataFrame(nqubits=4:17,
    qiskitX=benchmarks_qiskit["X"],
    qiskitH=benchmarks_qiskit["H"],
    qiskitT=benchmarks_qiskit["T"],
    qiskitCNOT=benchmarks_qiskit["CNOT"],
    qiskitToffoli=benchmarks_qiskit["Toffoli"])

CSV.write("qiskit.csv", data)

file = JSON.parsefile("projectq/.benchmarks/Darwin-CPython-3.7-64bit/0001_e5b779f8cec7ee91655814cb360026fad83667bd_20190703_230034_uncommited-changes.json")

benchmarks_projectq = Dict()
benchmarks_projectq["X"] = [each["stats"]["min"] * 1e9 for each in file["benchmarks"] if each["group"] == "X"]
benchmarks_projectq["H"] = [each["stats"]["min"] * 1e9 for each in file["benchmarks"] if each["group"] == "H"]
benchmarks_projectq["T"] = [each["stats"]["min"] * 1e9 for each in file["benchmarks"] if each["group"] == "T"]
benchmarks_projectq["CNOT"] = [each["stats"]["min"] * 1e9 for each in file["benchmarks"] if each["group"] == "CNOT"]
benchmarks_projectq["Toffoli"] = [each["stats"]["min"] * 1e9 for each in file["benchmarks"] if each["group"] == "Toffoli"]


data = DataFrame(nqubits=4:25,
X=benchmarks_projectq["X"],
H=benchmarks_projectq["H"],
T=benchmarks_projectq["T"],
CNOT=benchmarks_projectq["CNOT"],
Toffoli=benchmarks_projectq["Toffoli"])

CSV.write("projectq.csv", data)

@load "yao/shortlist.jld" benchmarks

benchmarks["X"]

benchmarks_yao = Dict()
benchmarks_yao["X"] = [each[2] for each in benchmarks["X"]]
benchmarks_yao["H"] = [each[2] for each in benchmarks["H"]]
benchmarks_yao["T"] = [each[2] for each in benchmarks["T"]]
benchmarks_yao["CNOT"] = [each[2] for each in benchmarks["CNOT"]]
benchmarks_yao["Toffoli"] = [each[2] for each in benchmarks["Toffoli"]]

data = DataFrame(nqubits=4:26,
X=benchmarks_yao["X"],
H=benchmarks_yao["H"],
T=benchmarks_yao["T"],
CNOT=benchmarks_yao["CNOT"],
Toffoli=benchmarks_yao["Toffoli"])

CSV.write("yao.csv", data)


file = JSON.parsefile("/Users/roger/.julia/dev/YaoBenchmarks/qcbm/0001_d8bf6f271550614cbf4e5f7f030cf145977f27d7_20190526_230620_uncommited-changes.json")
qcbm_benchmark["projectq"] = [each["stats"]["min"] * 1e9 for each in file["benchmarks"] if each["group"] == "qcbm"]

@load "/Users/roger/.julia/dev/YaoBenchmarks/qcbm.jld" qcbm

qcbm_benchmark["yao"] = qcbm

data = DataFrame(nqubits=4:14,
    yao=qcbm_benchmark["yao"][1:end-1],
    projectq=qcbm_benchmark["projectq"][1:end-1],
    cirq=qcbm_benchmark["cirq"],
    qiskit=qcbm_benchmark["qiskit"]
)

CSV.write("qcbm.csv", data)
